From: =?utf-8?b?Ik1hcmNvIFRyZXZpc2FuIChUcmV2acOxbyki?= <mail@3v1n0.net>
Date: Wed, 1 May 2024 04:20:42 +0200
Subject: shell-tray-icon: Trigger a relayout when the underneath xembeded
 size changes

(cherry picked from commit 8ebb71d9fdb8990e4e9731cc7662dbf704ce42a8)

Origin: https://gitlab.gnome.org/3v1n0/gnome-shell/-/commits/tray-offscreen-xwindows
Bug-Ubuntu: https://bugs.launchpad.net/bugs/2012388
---
 src/shell-tray-icon.c |  4 ++++
 src/tray/na-xembed.c  | 18 ++++++++++++++++++
 src/tray/na-xembed.h  |  1 +
 3 files changed, 23 insertions(+)

diff --git a/src/shell-tray-icon.c b/src/shell-tray-icon.c
index e1ff1cd..ddf40dd 100644
--- a/src/shell-tray-icon.c
+++ b/src/shell-tray-icon.c
@@ -265,6 +265,10 @@ shell_tray_icon_set_child (ShellTrayIcon *tray_icon,
                       "window-created",
                       G_CALLBACK (shell_tray_icon_window_created_cb),
                       tray_icon);
+
+  g_signal_connect_object (tray_child, "reconfigured",
+                           G_CALLBACK (clutter_actor_queue_relayout),
+                           tray_icon, G_CONNECT_SWAPPED);
 }
 
 /*
diff --git a/src/tray/na-xembed.c b/src/tray/na-xembed.c
index 447c1cb..873d193 100644
--- a/src/tray/na-xembed.c
+++ b/src/tray/na-xembed.c
@@ -91,6 +91,7 @@ G_DEFINE_TYPE_WITH_PRIVATE (NaXembed, na_xembed, G_TYPE_OBJECT)
 enum {
   PLUG_ADDED,
   PLUG_REMOVED,
+  RECONFIGURED,
   LAST_SIGNAL
 };
 
@@ -548,6 +549,15 @@ xembed_filter_func (MetaX11Display *x11_display,
           }
         break;
       }
+    case ConfigureNotify:
+      {
+        XConfigureEvent *xce = &xevent->xconfigure;
+
+        if (xce->window == priv->socket_window)
+          g_signal_emit (xembed, signals[RECONFIGURED], 0);
+
+        break;
+      }
     case DestroyNotify:
       {
         XDestroyWindowEvent *xdwe = &xevent->xdestroywindow;
@@ -742,6 +752,14 @@ na_xembed_class_init (NaXembedClass *klass)
                   NULL, NULL, NULL,
                   G_TYPE_NONE, 0);
 
+  signals[RECONFIGURED] =
+    g_signal_new ("reconfigured",
+                  G_OBJECT_CLASS_TYPE (klass),
+                  G_SIGNAL_RUN_LAST,
+                  G_STRUCT_OFFSET (NaXembedClass, reconfigured),
+                  NULL, NULL, NULL,
+                  G_TYPE_NONE, 0);
+
   props[PROP_X11_DISPLAY] =
     g_param_spec_object ("x11-display",
                          "x11-display",
diff --git a/src/tray/na-xembed.h b/src/tray/na-xembed.h
index 751da97..a5a4fae 100644
--- a/src/tray/na-xembed.h
+++ b/src/tray/na-xembed.h
@@ -36,6 +36,7 @@ struct _NaXembedClass
 
   void (* plug_added) (NaXembed *xembed);
   void (* plug_removed) (NaXembed *xembed);
+  void (* reconfigured) (NaXembed *xembed);
 };
 
 MetaX11Display * na_xembed_get_x11_display (NaXembed *xembed);
